### PR TITLE
Added support for sorting scalar data types

### PR DESCRIPTION
**Pull request by ribaraka**
Refs #457.

- improved regex comparison.
- fix `comment` bug.

### CODE REVIEW

**Review by seeforschauer — COMMENTED** [review]
LGTM, seems like need a call on this.
Can be there any errors added in test?

File: internal/types/compare.go [hunk 1]
```diff
@@ -255,3 +251,121 @@ func compareNumbers(a float64, b int64) CompareResult {
...
+	doubleNegativeZero
+	doubleDT
+	int32DT
+	int64DT
+)
+
+// defineNumberDataType define which number type has value and returns a sequence the type has.
+func defineNumberDataType(value any) numberDataTypeOrderResult {
+	switch value := value.(type) {
+	case int64:
```
Comment (w84thesun) [inline internal/types/compare.go:321]: Please rearrange cases in order with [code style](https://github.com/FerretDB/FerretDB/blob/main/CONTRIBUTING.md#code-style-and-conventions)

File: internal/types/compare.go [hunk 2]
```diff
@@ -255,3 +251,121 @@ func compareNumbers(a float64, b int64) CompareResult {
...
+	dateDataType
+	timestampDataType
+	regexDataType
+	maxKeyDataType
+)
+
+// defineDataType define which type has value and returns a sequence the type has.
+func defineDataType(value any) dataTypeOrderResult {
+	switch value := value.(type) {
+	case NullType:
```
Comment (w84thesun) [inline internal/types/compare.go:279]: Please rearrange cases in order with [code style](https://github.com/FerretDB/FerretDB/blob/main/CONTRIBUTING.md#code-style-and-conventions)

**Review by w84thesun — CHANGES_REQUESTED** [review]

File: internal/types/compare.go (hunk 1 above)
Comment (ribaraka) [reply internal/types/compare.go]: good catch! thank you! the cases have been rearrangeg.

File: internal/types/compare.go (hunk 2 above)
Comment (ribaraka) [reply internal/types/compare.go]: fixed.

**Comment by ribaraka** [comment]
> @ribaraka would you please re-run `task gen` locally and commit that change? Then dirty check will pass. Thank you.

done

File: internal/handlers/common/error.go [hunk 3]
```diff
@@ -41,7 +41,7 @@ const (
 	ErrNamespaceExists   = ErrorCode(48)    // NamespaceExists
 	ErrCommandNotFound   = ErrorCode(59)    // CommandNotFound
 	ErrNotImplemented    = ErrorCode(238)   // NotImplemented
-	ErrSortBadValue      = ErrorCode(15974) // Location15974
```
Comment (seeforschauer) [inline internal/handlers/common/error.go:44]: hm, it means that some test cases are missing.  Would you please add them?

Comment (ribaraka) [reply internal/handlers/common/error.go:44]: Could you please point to one?

Comment (w84thesun) [reply internal/handlers/common/error.go:44]: There is one test we can add for this error type. Could you please add this test in `TestQueryBadSortType`?
```go
		"BadSortTypeValue": {
			command: bson.D{
				{"find", collection.Name()},
				{"projection", bson.D{{"value", "some"}}},
				{"sort", bson.D{{"asc", "123"}}},
			},
			err: &mongo.CommandError{
				Code:    15974,
				Name:    "Location15974",
				Message: `Illegal key in $sort specification: asc: "123"`,
			},
		},
```

Comment (seeforschauer) [reply internal/handlers/common/error.go:44]: @w84thesun good example, thank you.

File: internal/types/compare.go [hunk 4]
```diff
@@ -34,13 +34,10 @@ type CompareResult int8
...
-	Less    CompareResult = -1 // <
-	Greater CompareResult = 1  // >
-
-	// TODO Remove once it is no longer needed.
-	// It should not be needed.
-	NotEqual CompareResult = 127 // !=
+	Equal        CompareResult = 0   // ==
+	Less         CompareResult = -1  // <
+	Greater      CompareResult = 1   // >
+	Incomparable CompareResult = 127 // !=
```
[the remark this answers is not in the document]
Comment (AlekSi) [reply internal/types/compare.go]: Unfortunately, `<>` is exactly the same as `!=` in SQL. From [this page](https://www.postgresql.org/docs/current/functions-comparison.html):
> <> is the standard SQL notation for “not equal”. != is an alias, which is converted to <> at a very early stage of parsing

After looking into [that Wikipedia page](https://en.wikipedia.org/wiki/Mathematical_operators_and_symbols_in_Unicode) for way too long, I propose to use `≹` (U+2279). Bonus point for that symbol looking just like crossed-out Z.

**Review by AlekSi — CHANGES_REQUESTED** [review]
> added support for storing negative zero

I added a "simple" integration test to see how MongoDB works with negative zero and noticed that our comparison function does not actually compare it correctly.
Please remove that from this PR for now.

File: internal/handlers/common/error.go (hunk 3 above)
Comment (ribaraka) [reply internal/handlers/common/error.go:44]: Thank you both for comments. Error case has been added. Error Code has been returned back, but slightly refactored.

**Comment by AlekSi** [comment]
Added #612 to track negative zero support.

File: internal/types/compare.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (ribaraka) [reply internal/types/compare.go]: :smile: agreed.

**Review by w84thesun — APPROVED** [review]

**Review by seeforschauer — APPROVED** [review]

File: internal/types/compare.go [hunk 5]
```diff
@@ -255,3 +251,123 @@ func compareNumbers(a float64, b int64) CompareResult {
...
+		return int32DT
+	case int64:
+		return int64DT
+	}
+
+	panic(fmt.Sprintf("defineNumberDataType: value cannot be defined, value is %v, data type of value is %T", value, value))
+}
+
+// CompareOrder defines the data type for the two values and compares them. When the types are equal, it compares using Compare.
+func CompareOrder(a, b any, order string) CompareResult {
```
Comment (AlekSi) [inline internal/types/compare.go:338]: We should use some other type for `order` instead of string, and use typed constants.

I guess we have to move `common.sortType` there and export it

File: internal/types/compare.go [hunk 6]
```diff
@@ -255,3 +251,123 @@ func compareNumbers(a float64, b int64) CompareResult {
 
 	return CompareResult(bigA.Cmp(bigB))
 }
+
+// dataTypeOrderResult represents the comparison order of data types.
+type dataTypeOrderResult uint8
+
+const (
+	_ dataTypeOrderResult = iota
+	minKeyDataType
```
[the remark this answers is not in the document]
Comment (AlekSi) [reply internal/types/compare.go:260]: I think yes; we need them for logging, debugging, error messages, etc.

File: internal/types/compare.go (hunk 5 above)
Comment (ribaraka) [reply internal/types/compare.go:338]: > We should use some other type for `order` instead of string...

What do you mean by that? Could you please give an example?

File: internal/types/compare.go [hunk 7]
```diff
@@ -255,3 +251,123 @@ func compareNumbers(a float64, b int64) CompareResult {
...
+	case int32:
+		return int32DT
+	case int64:
+		return int64DT
+	}
+
+	panic(fmt.Sprintf("defineNumberDataType: value cannot be defined, value is %v, data type of value is %T", value, value))
+}
+
+// CompareOrder defines the data type for the two values and compares them. When the types are equal, it compares using Compare.
```
[the remark this answers is not in the document]
Comment (AlekSi) [reply internal/types/compare.go:337]: My understanding is that we will use `Compare` for filtering and `CompareOrder` for sorting. Is that correct?

If yes, then `Compare`'s function documentation says "filtering or sorting" which is not correct: it should not mention sorting

[the remark this answers is not in the document]
Comment (ribaraka) [reply internal/types/compare.go:337]: When in `CompareOrder` types are equal, it then compares their values using `Compare`.

File: internal/types/compare.go (hunk 5 above)
Comment (ribaraka) [reply internal/types/compare.go:338]: Got it! Again, thank you for you time :smile:

Comment (ribaraka) [reply internal/types/compare.go:338]: Should I also move all those orders of data types (`dataTypeOrderResult`, `numberDataTypeOrderResult`) to common.sortType as well?

Comment (AlekSi) [reply internal/types/compare.go:338]: I think we should keep `CompareOrder` in `types` package. For that, we should move `common.sortType` to `types` package. To be able to use, we should export it.

I don't see why we need to move or export `dataTypeOrderResult` and `numberDataTypeOrderResult`

File: internal/types/compare.go [hunk 8]
```diff
@@ -255,3 +251,123 @@ func compareNumbers(a float64, b int64) CompareResult {
...
+		return regexDataType
+	case int32:
+		return numbersDataType
+	case Timestamp:
+		return timestampDataType
+	case int64:
+		return numbersDataType
+	}
+
+	panic(fmt.Sprintf("value cannot be defined, value is %v, data type of value is %T", value, value))
```
[the remark this answers is not in the document]
Comment (ribaraka) [reply internal/types/compare.go:306]: fixed.

File: integration/query_test.go [hunk 9]
```diff
@@ -168,6 +168,129 @@ func TestQuerySort(t *testing.T) {
...
+				"int64-zero",
+				"int32-min",
+				"int64-min",
+				"double-negative-infinity",
+				"double-nan",
+				"null",
+			},
+		},
+		"BadSortValue": {
+			sort: bson.D{{"value", 11}},
```
[the remark this answers is not in the document]
Comment (ribaraka) [reply integration/query_test.go:266]: The tests have been added.

[the remark this answers is not in the document]
Comment (ribaraka) [reply integration/query_test.go:266]: Do I need to add tests with other data types?

File: internal/types/compare.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (ribaraka) [reply internal/types/compare.go]: Added.

File: internal/types/compare.go (hunk 5 above)
Comment (ribaraka) [reply internal/types/compare.go]: migrated

File: integration/query_test.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (ribaraka) [reply integration/query_test.go:266]: @AlekSi  ^

[the remark this answers is not in the document]
Comment (seeforschauer) [reply integration/query_test.go:266]: @ribaraka which data types do you mean?

[the remark this answers is not in the document]
Comment (seeforschauer) [reply integration/query_test.go:266]: The PR title is `scalar data types`, so suppose no need for documents and arrays. 

What about strings. Do you mean to have some scalar type lists, and try to sort them with different options? Definitely worth it.

[the remark this answers is not in the document]
Comment (ribaraka) [reply integration/query_test.go:266]: > What about strings. Do you mean to have some scalar type lists, and try to sort them with different options? Definitely worth it.

I'd like to do it in another PR.

@AlekSi suggested to test sorting using BadValue as null and 0, now I've been suggesting adding a `BadValue` as another different data types.

File: internal/types/compare.go [hunk 10]
```diff
@@ -27,40 +27,37 @@ import (
...
 
-// Compare compares any BSON values in the same way as MongoDB does it for filtering or sorting.
+// Compare compares any BSON values in the same way as MongoDB does it for filtering.
 //
 // It converts types as needed; that may result in different types being equal.
 // For that reason, it typically should not be used in tests.
 //
 // Compare and contrast with test helpers in testutil package.
 func Compare(v1, v2 any) CompareResult {
-	if v1 == nil {
```
Comment (seeforschauer) [inline internal/types/compare.go:53]: why this change?  as for me previous version was more readable.

File: internal/types/compare.go [hunk 11]
```diff
@@ -255,3 +254,130 @@ func compareNumbers(a float64, b int64) CompareResult {
...
+	arrayDataType
+	binDataType
+	objectIdDataType
+	booleanDataType
+	dateDataType
+	timestampDataType
+	regexDataType
+)
+
+// defineDataType define which type has value and returns a sequence the type has.
```
Comment (seeforschauer) [inline internal/types/compare.go:277]:
```suggestion
// defineDataType returns a sequence for build-in type.
```

Comment (seeforschauer) [reply internal/types/compare.go:277]: would you be so kind to add a comment about Document and Array types?

File: internal/types/compare.go [hunk 12]
```diff
@@ -255,3 +254,130 @@ func compareNumbers(a float64, b int64) CompareResult {
...
+
+const (
+	_ numberDataTypeOrderResult = iota
+	doubleNegativeZero
+	doubleDT
+	int32DT
+	int64DT
+)
+
+// defineNumberDataType define which number type has value and returns a sequence the type has.
```
Comment (seeforschauer) [inline internal/types/compare.go:321]:
```suggestion
// defineNumberDataType returns a sequence for float64, int32 and int64 types.
```

File: internal/handlers/pg/msg_find.go [hunk 13]
```diff
@@ -92,13 +92,16 @@ func (h *Handler) MsgFind(ctx context.Context, msg *wire.OpMsg) (*wire.OpMsg, er
...
 		return nil, err
 	}
-	// comment in query, e.g. db.collection.find({$comment: "test"})
-	if sp.comment, err = common.GetOptionalParam(filter, "$comment", sp.comment); err != nil {
-		return nil, err
+	// get comment from query, e.g. db.collection.find({$comment: "test"})
+	if filter != nil {
+		if sp.comment, err = common.GetOptionalParam(filter, "$comment", sp.comment); err != nil {
+			return nil, err
+		}
```
Comment (AlekSi) [inline internal/handlers/pg/msg_find.go:104]: Should we do the same in findAndModify?

File: internal/types/compare.go [hunk 14]
```diff
@@ -255,3 +254,130 @@ func compareNumbers(a float64, b int64) CompareResult {
...
+	}
+
+	panic(fmt.Sprintf("defineNumberDataType: value cannot be defined, value is %v, data type of value is %T", value, value))
+}
+
+// SortType represents sort type for $sort aggregation.
+type SortType int8
+
+const (
+	Ascending  SortType = 1  // asc
```
Comment (seeforschauer) [inline internal/types/compare.go:342]: Would you be so kind to add comments here for exported variables?

File: internal/types/compare.go [hunk 15]
```diff
@@ -255,3 +254,130 @@ func compareNumbers(a float64, b int64) CompareResult {
...
+			return doubleNegativeZero
+		}
+		return doubleDT
+	case int32:
+		return int32DT
+	case int64:
+		return int64DT
+	}
+
+	panic(fmt.Sprintf("defineNumberDataType: value cannot be defined, value is %v, data type of value is %T", value, value))
```
Comment (seeforschauer) [inline internal/types/compare.go:335]: Should we panic here?

File: internal/types/compare.go [hunk 16]
```diff
@@ -255,3 +254,130 @@ func compareNumbers(a float64, b int64) CompareResult {
...
+type SortType int8
+
+const (
+	Ascending  SortType = 1  // asc
+	Descending SortType = -1 // desc
+)
+
+// CompareOrder defines the data type for the two values and compares them.
+// When the types are equal, it compares their values using Compare.
+func CompareOrder(a, b any, order SortType) CompareResult {
```
Comment (AlekSi) [inline internal/types/compare.go:348]: Let's extract that function and everything about ordering into compare_order.go

File: internal/types/compare.go [hunk 17]
```diff
@@ -255,3 +254,130 @@ func compareNumbers(a float64, b int64) CompareResult {
...
+	if a == nil || b == nil {
+		panic(fmt.Sprintf("CompareOrder: values should not be nil, a is %v, b is %v", a, b))
+	}
+
+	aType := defineDataType(a)
+	bType := defineDataType(b)
+	if aType == bType {
+		res := Compare(a, b)
+		if res == Equal && aType == numbersDataType {
```
Comment (seeforschauer) [inline internal/types/compare.go:357]: could you please remove here cyclomatic complexity?

File: internal/types/compare.go [hunk 18]
```diff
@@ -255,3 +254,130 @@ func compareNumbers(a float64, b int64) CompareResult {
...
+			bNumberType := defineNumberDataType(b)
+			switch {
+			case aNumberType < bNumberType && order == Ascending:
+				return Less
+			case aNumberType > bNumberType && order == Ascending:
+				return Greater
+			case aNumberType < bNumberType && order == Descending:
+				return Greater
+			case aNumberType > bNumberType && order == Descending:
+				return Less
```
Comment (seeforschauer) [inline internal/types/compare.go:368]: default is a must for switch cases, could you please add it here?

File: internal/types/compare.go [hunk 19]
```diff
@@ -255,3 +254,130 @@ func compareNumbers(a float64, b int64) CompareResult {
...
+
+const (
+	Ascending  SortType = 1  // asc
+	Descending SortType = -1 // desc
+)
+
+// CompareOrder defines the data type for the two values and compares them.
+// When the types are equal, it compares their values using Compare.
+func CompareOrder(a, b any, order SortType) CompareResult {
+	if a == nil || b == nil {
```
Comment (seeforschauer) [inline internal/types/compare.go:349]: separate checks will be nicier.

File: internal/types/compare.go [hunk 20]
```diff
@@ -255,3 +254,130 @@ func compareNumbers(a float64, b int64) CompareResult {
...
+			}
+		}
+
+		return res
+	}
+
+	switch {
+	case aType < bType:
+		return Less
+	case aType > bType:
```
Comment (seeforschauer) [inline internal/types/compare.go:378]: would you be so kind to add default here?

**Review by seeforschauer — CHANGES_REQUESTED** [review]

File: internal/types/compare.go [hunk 21]
```diff
@@ -255,3 +254,130 @@ func compareNumbers(a float64, b int64) CompareResult {
 
 	return CompareResult(bigA.Cmp(bigB))
 }
+
+// dataTypeOrderResult represents the comparison order of data types.
+type dataTypeOrderResult uint8
```
Comment (AlekSi) [inline internal/types/compare.go:259]: I wonder if we could come up with a better name…

@rumyantseva thoughts?

File: internal/types/compare.go [hunk 22]
```diff
@@ -255,3 +254,130 @@ func compareNumbers(a float64, b int64) CompareResult {
...
+	binDataType
+	objectIdDataType
+	booleanDataType
+	dateDataType
+	timestampDataType
+	regexDataType
+)
+
+// defineDataType define which type has value and returns a sequence the type has.
+func defineDataType(value any) dataTypeOrderResult {
```
Comment (rumyantseva) [inline internal/types/compare.go:278]: As @AlekSi mentioned, `detectDataType` would be a better name here.

File: internal/types/compare.go (hunk 21 above)
Comment (rumyantseva) [reply internal/types/compare.go:259]: Maybe one "pro" of dedicating a separate package is to decrease the cognitive load... Fewer lines in the file, simpler mental context... But still, it's rather a random idea.

File: internal/handlers/common/error.go [hunk 23]
```diff
@@ -42,7 +42,8 @@ const (
 	ErrCommandNotFound   = ErrorCode(59)    // CommandNotFound
 	ErrInvalidNamespace  = ErrorCode(73)    // InvalidNamespace
 	ErrNotImplemented    = ErrorCode(238)   // NotImplemented
-	ErrSortBadValue      = ErrorCode(15974) // Location15974
+	ErrLocation15974     = ErrorCode(15974) // Location15974
```
Comment (seeforschauer) [inline internal/handlers/common/error.go:45]: @ribaraka could you please give here some meaningful name? Thank you.

File: internal/types/compare.go (hunk 15 above)
Comment (seeforschauer) [reply internal/types/compare.go:335]: > However I've moved panic to `default` case.

sorry I don't see your changes, could you please verify you have pushed?

File: internal/handlers/pg/msg_find.go (hunk 13 above)
Comment (ribaraka) [reply internal/handlers/pg/msg_find.go:104]: Not neccesary do it same in this PR.
This piece of code has been refactored due to a bug. The bug occurs when we do not provide a filter and at the same time try to get a value from it by key.
I can add `comment` in this PR or another. What do you think is the right thing to do?

File: internal/types/compare.go (hunk 21 above)
Comment (ribaraka) [reply internal/types/compare.go:259]: > Something like `typeInComparisonOrder` looks pretty much similar to how it's already 

@rumyantseva, how about `sequenceDatatype`?

Comment (rumyantseva) [reply internal/types/compare.go:259]: @AlekSi I believe, as the main guard of code standards in this repo, you should be the decision-maker here ;)

File: internal/types/compare.go (hunk 10 above)
Comment (ribaraka) [reply internal/types/compare.go]: returned to its original form

File: internal/types/compare.go (hunk 11 above)
Comment (ribaraka) [reply internal/types/compare.go]: A comment for `detectType` has been changed.

A comment for Document and Array - do you mean like TODO?

File: internal/types/compare.go (hunk 12 above)
Comment (ribaraka) [reply internal/types/compare.go]: fixed

File: internal/types/compare.go (hunk 14 above)
Comment (ribaraka) [reply internal/types/compare.go]: added.

File: internal/types/compare.go (hunk 15 above)
Comment (ribaraka) [reply internal/types/compare.go]: ooh, sorry, I had changed it locally. I only notified you about that to make sure that panic can be set by default.
A commit with changes has been already pushed. Please review.

File: internal/types/compare.go (hunk 17 above)
Comment (ribaraka) [reply internal/types/compare.go]: fixed.

File: internal/types/compare.go (hunk 18 above)
Comment (ribaraka) [reply internal/types/compare.go]: fixed

File: internal/types/compare.go (hunk 20 above)
Comment (ribaraka) [reply internal/types/compare.go]: added

File: internal/types/compare.go (hunk 16 above)
Comment (ribaraka) [reply internal/types/compare.go]: done

File: internal/types/compare.go (hunk 22 above)
Comment (ribaraka) [reply internal/types/compare.go]: done

File: internal/handlers/common/error.go (hunk 23 above)
Comment (ribaraka) [reply internal/handlers/common/error.go]: sure! done.

File: internal/types/compare_order.go [hunk 24]
```diff
@@ -0,0 +1,162 @@
...
+		panic("CompareOrder: b is nil")
+	}
+
+	aType := detectDataType(a)
+	bType := detectDataType(b)
+	switch {
+	case aType == bType:
+		res := Compare(a, b)
+
+		if res == Equal && aType == numbersDataType {
```
[the remark this answers is not in the document]
Comment (ribaraka) [reply internal/types/compare_order.go]: fixed.

File: internal/types/compare.go (hunk 17 above)
Comment (ribaraka) [reply internal/types/compare.go]: @seeforschauer, did my changes helped or should I try and do more?

File: internal/types/compare_order.go (hunk 24 above)
[the remark this answers is not in the document]
Comment (ribaraka) [reply internal/types/compare_order.go:137]: the code block has minor changes a little bit.

File: internal/types/compare.go (hunk 11 above)
Comment (seeforschauer) [reply internal/types/compare.go]: I just didn't understand - the func was accepting `any` and the comment specified that the func is for build-in type, but `Document` and `Array`, which are in build-in type, were not processed in the function.

**Review by seeforschauer — COMMENTED** [review]
LGTM, and let's add a TODO with an issue links.

File: internal/types/compare.go (hunk 11 above)
Comment (ribaraka) [reply internal/types/compare.go]: I'm going to engage them (`Document` and `Array` types) in another PR.

File: internal/handlers/pg/msg_find.go (hunk 13 above)
Comment (seeforschauer) [reply internal/handlers/pg/msg_find.go:104]: @ribaraka I think if you are about to divide PR either you don't close the current issue and add some comments to it or create a new issue with the tas (depends on situation and description)

**Review by seeforschauer — DISMISSED** [review]
`findAndModify` sorting of scalar data types can be done in a separate PR.
I wrote a comment in the related issue, hope that helps if it's not ok,  create a new one.

File: internal/handlers/pg/msg_find.go (hunk 13 above)
Comment (ribaraka) [reply internal/handlers/pg/msg_find.go:104]: Good thought! The issue has been added.

File: integration/query_test.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (AlekSi) [reply integration/query_test.go:267]: #636

File: internal/handlers/common/sort.go [hunk 25]
```diff
@@ -141,13 +130,16 @@ func (ds *docsSorter) Less(i, j int) bool {
 	return ds.sorts[k](p, q)
 }
 
-// getSortType determines sortType from input sort value.
-func getSortType(value any) (sortType, error) {
+// getSortType determines SortType from input sort value.
+func getSortType(key, value any) (types.SortType, error) {
```
Comment (AlekSi) [inline internal/handlers/common/sort.go:134]: `key` is always `string`

File: internal/types/compare.go [hunk 26]
```diff
@@ -34,16 +34,13 @@ type CompareResult int8
 // Values match results of comparison functions such as bytes.Compare.
 // They do not match MongoDB `sort` values where 1 means ascending order and -1 means descending.
```
Comment (AlekSi) [inline internal/types/compare.go:35]: Let's mention `SortType` there

**Review by AlekSi — CHANGES_REQUESTED** [review]

File: internal/types/compare_order.go [hunk 27]
```diff
@@ -0,0 +1,167 @@
...
+	// Ascending is used for sort in ascending order.
+	Ascending SortType = 1
+
+	// Descending is used for sort in descending order.
+	Descending SortType = -1
+)
+
+// CompareOrder detects the data type for two values and compares them.
+// When the types are equal, it compares their values using Compare.
+func CompareOrder(a, b any, order SortType) CompareResult {
```
[the remark this answers is not in the document]
Comment (ribaraka) [reply internal/types/compare_order.go:120]: ascending

File: internal/types/compare_order.go [hunk 28]
```diff
@@ -0,0 +1,167 @@
...
+// compareTypeOrderResult represents the comparison order of data types.
+type compareTypeOrderResult uint8
+
+const (
+	_ compareTypeOrderResult = iota
+	nullDataType
+	nanDataType
+	numbersDataType
+	stringDataType
+	documentDataType // TODO: https://github.com/FerretDB/FerretDB/issues/457
```
[the remark this answers is not in the document]
Comment (ribaraka) [reply internal/types/compare_order.go]: fixed

File: internal/types/compare.go (hunk 26 above)
Comment (ribaraka) [reply internal/types/compare.go]: done

**Review by seeforschauer — APPROVED** [review]

**Review by AlekSi — APPROVED** [review]
